Data microservice integration

fgaribay-nearsoft opened the pull request:
This PR adds the data-layer microservice with the broker-based integration. This makes some modifications to the webapp (`management-app`) to include the connection to the broker and handling of validation. It also extracts repeated elements to a common library which also includes the broker configuration. Please advise on any issues you might find.

The commits are broadly grouped by:
- Structural set-up
- Data and broker configurations
- Microservice listeners creation
- Webapp integration
- Refactoring to common
- Simple validation

fgaribay-nearsoft committed:
Move Maven Wrapper to the parent project

fgaribay-nearsoft committed:
Update .gitignore

The same .gitignore would have been necessary for the new sub-module.

fgaribay-nearsoft committed:
Add POM for microservice submodule

Need to check which dependencies to extract to parent.

fgaribay-nearsoft committed:
Add skeleton for microservice module

fgaribay-nearsoft committed:
Add microservice configuration for broker

fgaribay-nearsoft committed:
Add datasource configuration

fgaribay-nearsoft committed:
Add model files

Will need to move this to a common submodule.

fgaribay-nearsoft committed:
Add stored procedure handling

fgaribay-nearsoft committed:
Add files for local deployment

fgaribay-nearsoft committed:
Add microservice listeners

fgaribay-nearsoft committed:
Update webapp to use direct exchanges

Currently using an RPC model. Need to discuss preferred architecture.

fgaribay-nearsoft committed:
Mark POJOs as serializable

Changed from List to ArrayList because it is serializable. Need to look into preferred solutions.
Will have to move to common library.

fgaribay-nearsoft committed:
Fix go to list after creating product

fgaribay-nearsoft committed:
Update tests to match refactoring

fgaribay-nearsoft committed:
Add common library sub-module

fgaribay-nearsoft committed:
Fix unnecessary dependency on parent POM

fgaribay-nearsoft committed:
Extract model into common library

fgaribay-nearsoft committed:
Use proper naming conventions

fgaribay-nearsoft committed:
Extract broker configuration to common library

fgaribay-nearsoft committed:
Reformat code in common

fgaribay-nearsoft committed:
Add simple validation

iuresti commented on common/pom.xml:
```diff
@@ -0,0 +1,47 @@
+        <dependency>
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+            <version>4.11</version>
```
use properties to specify versions

iuresti commented on management-microservice/src/main/java/com/nearsoft/fgaribay/mgmt/ProductRepositoryCustom.java:
```diff
@@ -0,0 +1,9 @@
+
+import java.util.List;
+
+public interface ProductRepositoryCustom {
```
Why you need ProductRepository and ProductRepositoryCustom?

carlos-hernandez replied:
I think it's required for Spring Data if you need to add additional functionality to the repositories it provides.

iuresti commented on common/src/main/java/com/nearsoft/fgaribay/mgmt/config/BrokerConfig.java:
```diff
@@ -0,0 +1,76 @@
+package com.nearsoft.fgaribay.mgmt.config;
+
+import org.springframework.amqp.core.*;
```
avoid wildcards.. I strongly recommend to auto format your code before commit. Doing that the IDE will expand your wildcards.

fgaribay-nearsoft replied:
Thanks, actually the IDE was the one to format it that way. IntelliJ uses by default wildcard imports when more than 5 classes from the same package are imported.

Should I disable it and update my code so that it matches the preferred style?

https://www.jetbrains.com/help/idea/creating-and-optimizing-imports.html#disable-wildcard-imports

iuresti commented on common/src/main/resources/common.properties:
```diff
@@ -0,0 +1,12 @@
+spring.rabbitmq.host=localhost
```
Why are you using this file instead of the default application.yml/application.properties ?

carlos-hernandez replied:
I'm nor sure if it would work, but you can create an `application-common.yml` file to use it as a Spring profile you can include in your other modules. The `@ConfigurationProperties` class that you have should work in the same way without needing to use the `@PropertySource` annotation.

fgaribay-nearsoft replied:
In the `common` library I am using a different file name so that it can be imported into other modules by using `@PropertySource("classpath:common.properties")`. This way, I have common properties defined in this module and they can be overriden by the corresponding `application.yml`.

The name `common.properties` also, from my perspective, makes it clearer that this is a shared properties file and not just used for the module it is stored in.

iuresti commented on common/src/main/java/com/nearsoft/fgaribay/mgmt/config/ProductProperties.java:
```diff
@@ -0,0 +1,112 @@
+@Validated
+@Configuration
+@ConfigurationProperties(prefix = "product")
+public class ProductProperties {
```
I recommend to put only one class per file

carlos-hernandez replied:
Agree, you can have all the static classes in its own file with a package scope.

iuresti commented on common/src/main/java/com/nearsoft/fgaribay/mgmt/config/ProductProperties.java:
```diff
@@ -0,0 +1,112 @@
+import javax.validation.constraints.NotNull;
+
+@Validated
+@Configuration
```
Do you really need `@Configuration` here?

iuresti commented on common/src/main/java/com/nearsoft/fgaribay/mgmt/model/Product.java:
```diff
@@ -0,0 +1,51 @@
+import java.io.Serializable;
+
+public class Product implements Serializable {
+  @Range(min = 1, max = 65535)
```
Why the range? is not this value auto generated from your database?

fgaribay-nearsoft replied:
As the business rules for the task are undefined, I created the following scenario:
* The system must allow the following legacy requirements:
  * Some outdated equipment will make use of the product information (e.g. 2 bytes in a barcode store the product ID).
  * The IDs cannot be autogenerated because control is needed by business over the IDs.

The reason for this is simply to include some form of validation. But there are different things I could validate if preferred.

Do you suggest I change the mock requirements to match a scenario more similar to SKT?

iuresti commented on management-app/src/main/java/com/nearsoft/fgaribay/mgmt/BrokerConnector.java:
```diff
@@ -8,26 +8,28 @@
     }
 
-    public List<Product> retrieveProductList() {
+    public ArrayList<Product> retrieveProductList() {
```
Why did you decide this should be an ArrayList instead of a List?

fgaribay-nearsoft replied:
Thanks, I originally thought it was necessary for serialization. I have fixed it now to use List.

iuresti commented on management-microservice/src/main/java/com/nearsoft/fgaribay/mgmt/ProductRepositoryImpl.java:
```diff
@@ -0,0 +1,26 @@
+    createProductQuery.setParameter("p_id", id);
+    createProductQuery.setParameter("p_name", name);
+    createProductQuery.setParameter("p_description", description);
+    createProductQuery.execute();
```
How the user of this method know if the product was actually inserted

iuresti commented on management-microservice/src/main/java/com/nearsoft/fgaribay/mgmt/microservice/ProductManagementService.java:
```diff
@@ -0,0 +1,46 @@
+    boolean error = false;
+    String errorMessage = "";
+    UUID uuid = request.getId();
+    Product product = request.getProducts().get(0);
```
Why are you sending a list if at the end you only read the first value?

fgaribay-nearsoft replied:
I use the same ProductRequest and ProductResponse for both current microservice operations (create one product and list all) and to support multiple products, I use a list. I went with this decision because other operations might be required in the future, such as creating multiple products.

The specific operation of creating a single product, means that there must be exactly one element in the list.

This is probably and the best solution might be to change the operation from _create a product_ to _create products_.

However, the requirements are to create one product. As such, **would you suggest that I use a different request type for creating a single product or that I add validations to ensure that exactly one product is received?**

fgaribay-nearsoft replied:
Hi Ivan! Talking to my mentors, we decided that it might look better to use generics here, so instead of having a ProductRequest, we would have a ServiceRequest with a generic data field.

If it's ok with you, I would like to leave that for the next pull request, which I will try to submit today.

iuresti commented on management-microservice/src/main/resources/application.yml:
```diff
@@ -0,0 +1,10 @@
+spring:
+  datasource:
+    url: jdbc:mysql://localhost/skt
+    username: root
```
don't store usernames or passwords in your repository ... you can use placeholders instead

carlos-hernandez commented on pom.xml:
```diff
@@ -22,6 +22,8 @@
 
```
If you run `mvn clean install` for this project the following error appears in the console:

`Error assembling WAR: webxml attribute is required (or pre-existing WEB-INF/web.xml if executing in update mode)`

Maybe you need to add some configurations for the `maven-war-plugin` to avoid this.

Also, use Maven properties to specify the version for all your dependencies and plugins, like the Spring Boot Maven plugin. Specifying a plugin version contributes to the stability of your builds.

carlos-hernandez commented on management-app/pom.xml:
```diff
@@ -50,6 +50,12 @@
+        <dependency>
+            <groupId>com.nearsoft.fgaribay</groupId>
+            <artifactId>common</artifactId>
+            <version>1.0-SNAPSHOT</version>
```
Remove `<version>0.0.1-SNAPSHOT</version>` at the start of this file and you can specify the version for this dependency like this:

```xml
<dependency>
    <groupId>com.nearsoft.fgaribay</groupId>
    <artifactId>common</artifactId>
    <version>${project.version}</version>
</dependency>
```

carlos-hernandez commented on management-microservice/pom.xml:
```diff
@@ -0,0 +1,61 @@
+        <dependency>
+            <groupId>com.nearsoft.fgaribay</groupId>
+            <artifactId>common</artifactId>
+            <version>1.0-SNAPSHOT</version>
```
See comment in `management-app/pom.xml` file

carlos-hernandez commented on common/src/main/java/com/nearsoft/fgaribay/mgmt/config/BrokerConfig.java:
```diff
@@ -0,0 +1,76 @@
+
+@Configuration
+@PropertySource("classpath:common.properties")
+@ComponentScan("com.nearsoft.fgaribay.mgmt.config")
```
By default, `@ComponentScan` uses the same package of the class that declares this annotation. You can safely remove the hardcoded package.

carlos-hernandez commented on common/src/main/java/com/nearsoft/fgaribay/mgmt/model/ProductRequest.java:
```diff
@@ -0,0 +1,48 @@
+public class ProductRequest implements Serializable {
+  private UUID id;
+  private String operation;
+  private ArrayList<Product> products;
```
Use the `List` interface instead of `ArrayList`. See if you are doing the same for other classes in your project.

carlos-hernandez commented on management-app/src/main/java/com/nearsoft/fgaribay/mgmt/BrokerConnector.java:
```diff
@@ -8,26 +8,28 @@
+    private ProductProperties.Exchanges exchangeProperties;
+    private ProductProperties.RoutingKeys routingKeyProperties;
 
     @Autowired
```
`@Autowired` is not needed here since you are using constructor dependency injection. Check it you are using this annotation in other beans and remove it.

carlos-hernandez commented on management-app/src/main/java/com/nearsoft/fgaribay/mgmt/ManagementApplication.java:
```diff
@@ -1,9 +1,14 @@
+import org.springframework.context.annotation.PropertySource;
 
 @SpringBootApplication
+@PropertySource("classpath:common.properties")
```
See the comment in `common/src/main/resources/common.properties` file and try using a Spring profile for this common settings.

carlos-hernandez commented on management-app/src/main/java/com/nearsoft/fgaribay/mgmt/ManagementController.java:
```diff
@@ -34,8 +35,11 @@ public String greetingForm(Model model) {
     }
 
     @RequestMapping(value = "/create", method = RequestMethod.POST)
```
Try replacing this mapping annotations with `@GetMapping` or `@PostMapping` as needed.

carlos-hernandez commented on management-microservice/src/main/java/com/nearsoft/fgaribay/mgmt/microservice/ManagementMicroserviceApplication.java:
```diff
@@ -0,0 +1,21 @@
+import org.springframework.data.jpa.repository.config.EnableJpaRepositories;
+
+@SpringBootApplication
+@EntityScan(basePackages = {"com.nearsoft.fgaribay.mgmt.model"})
```
Use `basePackageClasses` instead of `basePackage` to avoid harcoding package names.

carlos-hernandez commented on management-microservice/src/main/java/com/nearsoft/fgaribay/mgmt/ProductRepositoryCustom.java:
```diff
@@ -0,0 +1,9 @@
+import java.util.List;
+
+public interface ProductRepositoryCustom {
+  List getAllProducts();
```
Why you don't specify the result type as `List<ProductEntity>`?

carlos-hernandez commented on management-microservice/src/main/java/com/nearsoft/fgaribay/mgmt/microservice/ProductManagementService.java:
```diff
@@ -0,0 +1,46 @@
+  @RabbitListener(queues = "${product.queues.list-name}")
+  public ProductResponse listProducts(ProductRequest request) {
+
+    ArrayList<Product> products = new ArrayList<>(productRepository.getAllProducts());
```
Using `List` you avoid to create a new array list with the same results you get from the repository. With that, the repository can define which `List` implementation uses to give you the data and the objects using the result won't care at all of the final type since it will be always a `List`.

carlos-hernandez commented on management-microservice/src/main/java/com/nearsoft/fgaribay/mgmt/microservice/ProductManagementService.java:
```diff
@@ -0,0 +1,46 @@
+    Product product = request.getProducts().get(0);
+    productRepository.createProduct(product.getId(), product.getName(), product.getDescription());
+
+    return new ProductResponse("create", error, errorMessage, request.getProducts(), uuid);
```
Be aware of the scenario in which you cannot save the product and change the response according to this. Right now you're always telling your client that there was no error saving the product, which is not always true.

carlos-hernandez commented on management-microservice/src/main/java/com/nearsoft/fgaribay/mgmt/model/ProductEntity.java:
```diff
@@ -0,0 +1,67 @@
+      name = "createProduct",
+      procedureName = "create_product",
+      parameters = {
+        @StoredProcedureParameter(name = "p_id", mode = ParameterMode.IN, type = Long.class),
```
`mode` is always `ParameterMode.IN` by default in this annotation.

Nitpick: Be consistent in the way you format your annotations XD

fgaribay-nearsoft replied:
About the formatting: Right? I had the same qualm with that, but the default code-style formatter disagreed. I'm changing it manually to match the other annotations and I'll see if I can find the right code-style setting.

fgaribay-nearsoft committed:
Separate type-safe property classes

fgaribay-nearsoft committed:
Remove unused Maven dependencies

fgaribay-nearsoft committed:
Avoid using wildcard imports

fgaribay-nearsoft committed:
Use List interface instead of ArrayList

fgaribay-nearsoft committed:
Remove credentials

Use placeholders instead

fgaribay-nearsoft committed:
Fix 'missing webxml' error

fgaribay-nearsoft committed:
Refactor POM files

- Move versions to properties
- Specify plugin versions
- Move plugins from parent to corresponding projects
- Reformat code

fgaribay-nearsoft committed:
Remove unnecessary Autowired annotations

fgaribay-nearsoft committed:
Use GetMapping and PostMapping

fgaribay-nearsoft committed:
Remove unused import

Change missed on previous commit.

fgaribay-nearsoft committed:
Remove unnecessary ProductRepositoryCustom interface

fgaribay-nearsoft committed:
Use basePackageClasses instead of basePackage

To avoid hardcoding package names.

fgaribay-nearsoft committed:
Use List<Product> for the ProductRepository

fgaribay-nearsoft committed:
Remove unnecessary parameter fom StoredProcedureParameter

fgaribay-nearsoft committed:
Add stored procedure error handling

fgaribay-nearsoft committed:
Remove hardcoded package in ComponentScan

fgaribay-nearsoft committed:
Use Spring profile for common properties

fgaribay-nearsoft committed:
Use Spring profile for common properties (missed file)

Continues previous commit. Missed file ManagementApplication.java

fgaribay-nearsoft committed:
Rename exception for clarity

fgaribay-nearsoft committed:
Add price to product model

To simulate a more real product and include other validations.

fgaribay-nearsoft committed:
Add messages to Product validation

This also makes the product ID a boxed Long to properly handle validation.

jass-holguin commented on management-microservice/src/main/java/com/nearsoft/fgaribay/mgmt/ProductRepository.java:
```diff
@@ -0,0 +1,14 @@
+import java.math.BigDecimal;
+import java.util.List;
+
+public interface ProductRepository extends Repository<ProductEntity, Long> {
```
If you're implementing manually this repository interface do you really have to extend from org.springframework.data.repository.Repository? Would it make any difference not to?

fgaribay-nearsoft replied:
Thanks for approving. I will use the cleaner JPA way that you mentioned on the call, and it should be there on the next pull request.
